hashes/src/sha256d.rs

```diff
@@ -13,10 +13,30 @@ crate::internal_macros::hash_type! {
     "Output of the SHA256d hash function."
 }
 
-type HashEngine = sha256::HashEngine;
+/// Engine to compute SHA256d hash function.
+#[derive(Clone)]
+pub struct HashEngine(sha256::HashEngine);
 
-fn from_engine(e: sha256::HashEngine) -> Hash {
-    let sha2 = sha256::Hash::from_engine(e);
+impl HashEngine {
+    /// Creates a new SHA256d hash engine.
+    pub const fn new() -> Self { Self(sha256::HashEngine::new()) }
+}
+
+impl Default for HashEngine {
+    fn default() -> Self { Self::new() }
+}
+
+impl crate::HashEngine for HashEngine {
+    type MidState = sha256::Midstate;
```

Comment:
Should we have a new type here as well?

Reply:
I thought that all the midstate stuff is a bit up in the air still so I just put it like this for now. Happy to take direction on it though, @apoelstra?

Reply:
Yeah, that's why I'm asking and not claiming it should be that way. :)

Reply:
Cool, thanks man. Thanks for your reviews today as well!

Reply:
I could really go either way on this, conceptually. But I think that if we're gonna separate the engine types we ought to also separate the midstates.

Reply:
But I'm also fine leaving it as-is for now. We can address it when we do https://github.com/rust-bitcoin/rust-bitcoin/issues/2918 (another "easy", or at least localized, `hashes` API change).

Reply:
Now I'm thinking having midstate on sha256t is really strange. Not strictly wrong but if you're using it maybe you meant to use sh256 midstate or you should've used our trait. So I suspect not all hashes should even support it.

Reply:
Yes, agreed. I think that probably only sha256 should. Easy to add to others later, even post-1.0.

But I also think this is a question for #2918, not this PR.